File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserGroupsControllerTest.java
package uk.gov.justice.digital.hmpps.oauth2server.resource.api;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.Group;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User;
import uk.gov.justice.digital.hmpps.oauth2server.maintain.AuthUserGroupService;
import uk.gov.justice.digital.hmpps.oauth2server.maintain.AuthUserGroupService.AuthUserGroupException;
import uk.gov.justice.digital.hmpps.oauth2server.maintain.AuthUserGroupService.AuthUserGroupExistsException;
import uk.gov.justice.digital.hmpps.oauth2server.maintain.AuthUserService;
import uk.gov.justice.digital.hmpps.oauth2server.model.AuthUserGroup;
import uk.gov.justice.digital.hmpps.oauth2server.model.ErrorDetail;

import java.security.Principal;
import java.util.List;
import java.util.Optional;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class AuthUserGroupsControllerTest {
    private final Principal principal = new UsernamePasswordAuthenticationToken("bob", "pass");

    @Mock
    private AuthUserService authUserService;
    @Mock
    private AuthUserGroupService authUserGroupService;

    private AuthUserGroupsController authUserGroupsController;

    @Before
    public void setUp() {
        authUserGroupsController = new AuthUserGroupsController(authUserService, authUserGroupService);
    }

    @Test
    public void groups_userNotFound() {
        final var responseEntity = authUserGroupsController.groups("bob");
        assertThat(responseEntity.getStatusCodeValue()).isEqualTo(404);
        assertThat(responseEntity.getBody()).isEqualTo(new ErrorDetail("Not Found", "Account for username bob not found", "username"));
    }

    @Test
    public void groups_success() {
        final var group1 = new Group("FRED", "desc");
        final var group2 = new Group("GLOBAL_SEARCH", "desc2");
        when(authUserGroupService.getAuthUserGroups(anyString())).thenReturn(Optional.of(Set.of(group1, group2)));
        final var responseEntity = authUserGroupsController.groups("joe");
        assertThat(responseEntity.getStatusCodeValue()).isEqualTo(200);
        //noinspection unchecked
        assertThat(((List) responseEntity.getBody())).containsOnly(new AuthUserGroup(group1), new AuthUserGroup(group2));
    }

    @Test
    public void addGroup_userNotFound() {
        final var responseEntity = authUserGroupsController.addGroup("bob", "group", principal);
        assertThat(responseEntity.getStatusCodeValue()).isEqualTo(404);
        assertThat(responseEntity.getBody()).isEqualTo(new ErrorDetail("Not Found", "Account for username bob not found", "username"));
    }

    @Test
    public void addGroup_success() throws AuthUserGroupException {
        when(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(getAuthUser()));
        final var responseEntity = authUserGroupsController.addGroup("someuser", "group", principal);
        assertThat(responseEntity.getStatusCodeValue()).isEqualTo(204);
        verify(authUserGroupService).addGroup("USER", "group", "bob");
    }

    @Test
    public void addGroup_conflict() throws AuthUserGroupException {
        when(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(getAuthUser()));
        doThrow(new AuthUserGroupExistsException()).when(authUserGroupService).addGroup(anyString(), anyString(), anyString());

        final var responseEntity = authUserGroupsController.addGroup("someuser", "joe", principal);
        assertThat(responseEntity.getStatusCodeValue()).isEqualTo(409);
    }

    @Test
    public void addGroup_validation() throws AuthUserGroupException {
        when(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(getAuthUser()));

        doThrow(new AuthUserGroupException("group", "error")).when(authUserGroupService).addGroup(anyString(), anyString(), anyString());
        final var responseEntity = authUserGroupsController.addGroup("someuser", "harry", principal);
        assertThat(responseEntity.getStatusCodeValue()).isEqualTo(400);
        assertThat(responseEntity.getBody()).isEqualTo(new ErrorDetail("group.error", "group failed validation", "group"));
    }

    @Test
    public void removeGroup_userNotFound() {
        final var responseEntity = authUserGroupsController.removeGroup("bob", "group", principal);
        assertThat(responseEntity.getStatusCodeValue()).isEqualTo(404);
        assertThat(responseEntity.getBody()).isEqualTo(new ErrorDetail("Not Found", "Account for username bob not found", "username"));
    }

    @Test
    public void removeGroup_success() throws AuthUserGroupException {
        when(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(getAuthUser()));
        final var responseEntity = authUserGroupsController.removeGroup("someuser", "joe", principal);
        assertThat(responseEntity.getStatusCodeValue()).isEqualTo(204);
        verify(authUserGroupService).removeGroup("USER", "joe", "bob");
    }

    @Test
    public void removeGroup_groupMissing() throws AuthUserGroupException {
        when(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(getAuthUser()));
        doThrow(new AuthUserGroupException("group", "error")).when(authUserGroupService).removeGroup(anyString(), anyString(), anyString());

        final var responseEntity = authUserGroupsController.removeGroup("someuser", "harry", principal);
        assertThat(responseEntity.getStatusCodeValue()).isEqualTo(400);
    }

    private User getAuthUser() {
        final var user = User.builder().username("USER").email("email").verified(true).build();
        user.setGroups(Set.of(new Group("GLOBAL_SEARCH", "desc2"), new Group("FRED", "desc")));
        return user;
    }
}

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserGroupsControllerTest.kt
package uk.gov.justice.digital.hmpps.oauth2server.resource.api

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.doThrow
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.userdetails.UsernameNotFoundException
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.ChildGroup
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.Group
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserHelper.Companion.createSampleUser
import uk.gov.justice.digital.hmpps.oauth2server.maintain.AuthUserGroupService
import uk.gov.justice.digital.hmpps.oauth2server.maintain.AuthUserGroupService.AuthUserGroupException
import uk.gov.justice.digital.hmpps.oauth2server.maintain.AuthUserGroupService.AuthUserGroupExistsException
import uk.gov.justice.digital.hmpps.oauth2server.maintain.AuthUserService
import uk.gov.justice.digital.hmpps.oauth2server.model.AuthUserGroup
import uk.gov.justice.digital.hmpps.oauth2server.model.ErrorDetail
import java.security.Principal
import java.util.Optional

class AuthUserGroupsControllerTest {
  private val principal: Principal = UsernamePasswordAuthenticationToken("bob", "pass")
  private val authUserService: AuthUserService = mock()
  private val authUserGroupService: AuthUserGroupService = mock()
  private val authUserGroupsController = AuthUserGroupsController(authUserService, authUserGroupService)

  @Nested
  inner class Groups {
    @Test
    fun `groups userNotFound`() {
      whenever(authUserGroupService.getAuthUserGroups(anyString())).thenReturn(null)
      assertThatThrownBy { authUserGroupsController.groups("bob") }
        .isInstanceOf(UsernameNotFoundException::class.java)
    }

    @Test
    fun `groups no children`() {
      val group1 = Group("FRED", "desc")
      val group2 = Group("GLOBAL_SEARCH", "desc2")
      whenever(authUserGroupService.getAuthUserGroups(anyString())).thenReturn(setOf(group1, group2))
      val responseEntity = authUserGroupsController.groups(username = "joe", children = false)
      assertThat(responseEntity).containsOnly(AuthUserGroup(group1), AuthUserGroup(group2))
    }

    @Test
    fun `groups default children`() {
      val group1 = Group("FRED", "desc")
      val group2 = Group("GLOBAL_SEARCH", "desc2")
      val childGroup = ChildGroup("CHILD_1", "child 1")
      group2.children.add(childGroup)
      whenever(authUserGroupService.getAuthUserGroups(anyString())).thenReturn(setOf(group1, group2))
      val responseEntity = authUserGroupsController.groups(username = "joe")
      assertThat(responseEntity).containsOnly(AuthUserGroup("FRED", "desc"), AuthUserGroup("CHILD_1", "child 1"))
    }

    @Test
    fun `groups with children requested`() {
      val group1 = Group("FRED", "desc")
      val group2 = Group("GLOBAL_SEARCH", "desc2")
      val childGroup = ChildGroup("CHILD_1", "child 1")
      group2.children.add(childGroup)
      whenever(authUserGroupService.getAuthUserGroups(anyString())).thenReturn(setOf(group1, group2))
      val responseEntity = authUserGroupsController.groups(username = "joe")
      assertThat(responseEntity).containsOnly(AuthUserGroup("FRED", "desc"), AuthUserGroup("CHILD_1", "child 1"))
    }
  }

  @Test
  fun addGroup_userNotFound() {
    val responseEntity = authUserGroupsController.addGroup("bob", "group", principal)
    assertThat(responseEntity.statusCodeValue).isEqualTo(404)
    assertThat(responseEntity.body).isEqualTo(
      ErrorDetail(
        "Not Found",
        "Account for username bob not found",
        "username"
      )
    )
  }

  @Test
  fun addGroup_success() {
    whenever(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(authUser))
    val responseEntity = authUserGroupsController.addGroup("someuser", "group", principal)
    assertThat(responseEntity.statusCodeValue).isEqualTo(204)
    verify(authUserGroupService).addGroup("USER", "group", "bob")
  }

  @Test
  fun addGroup_conflict() {
    whenever(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(authUser))
    doThrow(AuthUserGroupExistsException()).whenever(authUserGroupService)
      .addGroup(anyString(), anyString(), anyString())
    val responseEntity = authUserGroupsController.addGroup("someuser", "joe", principal)
    assertThat(responseEntity.statusCodeValue).isEqualTo(409)
  }

  @Test
  fun addGroup_validation() {
    whenever(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(authUser))
    doThrow(AuthUserGroupException("group", "error")).whenever(authUserGroupService)
      .addGroup(anyString(), anyString(), anyString())
    val responseEntity = authUserGroupsController.addGroup("someuser", "harry", principal)
    assertThat(responseEntity.statusCodeValue).isEqualTo(400)
    assertThat(responseEntity.body).isEqualTo(ErrorDetail("group.error", "group failed validation", "group"))
  }

  @Test
  fun removeGroup_userNotFound() {
    val responseEntity = authUserGroupsController.removeGroup("bob", "group", principal)
    assertThat(responseEntity.statusCodeValue).isEqualTo(404)
    assertThat(responseEntity.body).isEqualTo(
      ErrorDetail(
        "Not Found",
        "Account for username bob not found",
        "username"
      )
    )
  }

  @Test
  fun removeGroup_success() {
    whenever(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(authUser))
    val responseEntity = authUserGroupsController.removeGroup("someuser", "joe", principal)
    assertThat(responseEntity.statusCodeValue).isEqualTo(204)
    verify(authUserGroupService).removeGroup("USER", "joe", "bob")
  }

  @Test
  fun removeGroup_groupMissing() {
    whenever(authUserService.getAuthUserByUsername(anyString())).thenReturn(Optional.of(authUser))
    doThrow(AuthUserGroupException("group", "error")).whenever(authUserGroupService)
      .removeGroup(anyString(), anyString(), anyString())
    val responseEntity = authUserGroupsController.removeGroup("someuser", "harry", principal)
    assertThat(responseEntity.statusCodeValue).isEqualTo(400)
  }

  private val authUser: User
    get() {
      return createSampleUser(
        username = "USER",
        email = "email",
        verified = true,
        groups = setOf(Group("GLOBAL_SEARCH", "desc2"), Group("FRED", "desc"))
      )
    }
}
